common4j/src/main/com/microsoft/identity/common/java/telemetry/Telemetry.java

```diff
@@ -0,0 +1,334 @@
+
+/**
+ * A singleton class for logging Telemetry.
+ * Must be instantiated via {@link Telemetry.Builder} before use.
+ */
+public class Telemetry {
+    private final static String TAG = Telemetry.class.getSimpleName();
+    private static volatile Telemetry sTelemetryInstance = null;
+
+    // Suppressing rawtype warnings due to the generic type ITelemetryObserver
+    @SuppressWarnings(WarningType.rawtype_warning)
+    private static Queue<ITelemetryObserver> mObservers;
+
+    private Queue<Map<String, String>> mTelemetryRawDataMap;
+    private TelemetryConfiguration mDefaultConfiguration;
+    private AbstractTelemetryContext mTelemetryContext;
+    private boolean mIsDebugging;
+
+    //if the telemetry configuration is not set
+    private final boolean mIsTelemetryEnabled;
+
+    protected Telemetry() {
+        // Added for backcompat (with android-common code).
+        // Once that Telemetry class is removed. Remove this too!
+        mIsTelemetryEnabled = false;
+    }
+
+    private Telemetry(final Builder builder) {
+        if (builder == null
+                || builder.mTelemetryContext == null
+                || builder.mDefaultConfiguration == null) {
+            //we do not want to throw exception for telemetry.
+            Logger.warn(
+                    TAG,
+                    "Telemetry is disabled because the Telemetry context or configuration is null"
+            );
+            mIsTelemetryEnabled = false;
+        } else {
+            mIsTelemetryEnabled = true;
+            mDefaultConfiguration = builder.mDefaultConfiguration;
+            mTelemetryContext = builder.mTelemetryContext;
+            mIsDebugging = builder.mIsDebugging;
+            mTelemetryRawDataMap = new ConcurrentLinkedQueue<>();
+        }
+    }
+
+    /**
+     * Prepares instance using builder.
+     **/
+    private synchronized static Telemetry prepareInstance(Builder builder) {
+        sTelemetryInstance = new Telemetry(builder);
+
+        return sTelemetryInstance;
+    }
+
+    /**
+     * This is for getting instance of Telemetry
+     **/
+    public synchronized static Telemetry getInstance() {
+        // If sTelemetryInstance is not initialized, telemetry will be disabled.
+        if (sTelemetryInstance == null) {
+            new Builder().build();
+        }
+
+        return sTelemetryInstance;
+    }
+
+    private Queue<Map<String, String>> getRequestMap() {
+        return mTelemetryRawDataMap;
+    }
+
+    /**
+     * Register the observer to upload the telemetry data.
+     *
+     * @param observer ITelemetryObserver
+     */
+    public void addObserver(@SuppressWarnings(WarningType.rawtype_warning) final ITelemetryObserver observer) {
+        if (null == observer) {
+            throw new IllegalArgumentException("Telemetry Observer instance cannot be null");
+        }
+
+        // check to make sure we're not already dispatching elsewhere
+        if (null == mObservers) {
```

Comment:
This looks like a potential race condition - suggestion synchronizing the method or having `mObservers` default to being initialized